### DISCUSSION TITLE
Day 3 has arrived

### DISCUSSION THREAD

**Discussion by CarolineMorton** [CarolineMorton/AdventOfCode_WIR#4] [Ideas]
🎄🎄 **Day 3 has arrived**

Day 3 puzzle is here and it looks to be a complicated little beasty. We have to get useful content out of a "corrupted" string. This will put our knowledge of strings (String vs str) to good use in Rust!

See you in the comments!

Caroline

**Comment by hollygrimm** [comment]
My [day 3 solution](https://github.com/hollygrimm/AdventOfCode_WIR/tree/main/day_03), refactored into separate files, ala @CarolineMorton 😃

**Reply by CarolineMorton** [reply]
Hey, this is great to see! I love how you can use the `error.rs` file in all the projects now. Great use of both docstrings and tests per file. Awesome work!!

**Reply by CarolineMorton** [reply]
One minor comment (hope you don't mind 🙏):

`Lazy` from `once_cell::sync::Lazy` has now been stabilised and put into standard rust via `std::sync::LazyLock`. See [docs](https://doc.rust-lang.org/beta/std/sync/struct.LazyLock.html). Recent change - i think 1.80 maybe. I only know this because I had to change every single Lazy ito LazyLock n the entire Async Rust book as the change happened during writing 😬

**Reply by hollygrimm** [reply]
I appreciate any comments or suggestions, Caroline! I've [updated my code](https://github.com/hollygrimm/AdventOfCode_WIR/commit/53b95ead2d72c03e70b552f9cb7fc6f4d00b725f), thank you.

**Comment by em-baggie** [comment]
My day 3 [solution](https://github.com/em-baggie/AdventOfCode_WIR/tree/main/day_03)

EDIT: refactored solution into modules :)

**Reply by CarolineMorton** [reply]
Nice @em-baggie Let me know if you want a review

**Reply by em-baggie** [reply]
> Nice @em-baggie Let me know if you want a review

Yes please!! Any comments greatly appreciated 🙏 (still getting my head around the basics of rust)

**Reply by CarolineMorton** [reply]
### Great Work on the Code! 🌟

Well done for this! There are lots of good points ⭐ including the splitting of logic into files, and then using these imports in your main. It keeps the `main.rs` file nice and clear. 

### Things to Consider for Improvement ✨

#### 1. **Result vs expect**
I can see that you have used Result here - meaning that the output of this function should be either a String or an Error. 
```rust
use std::fs;

pub fn parse_input(file_path: &str) -> Result<String, std::io::Error> { 
    let file = fs::read_to_string(file_path).expect("Unable to read file"); 
    Ok(file)
}
```
The issue is that using `expect()` shortcuts this logic, by essentially stopping the program. The purpose of returning a Result is to allow the caller to decide how to handle an error. By using expect, you're making the decision (to panic), which negates the value of returning a Result. 

A way to rewrite this:
```rust
use std::fs;

pub fn parse_input(file_path: &str) -> Result<String, std::io::Error> {
    fs::read_to_string(file_path)
}
```
`fs::read_to_string` already returns a Result<String, std::io::Error> so we can use this directly, and any error will bubble up to our `prase_input` fn. 

#### 2. Add doc strings
I think it would be useful to add in some docstrings. It is a helpful practice to get into. I always think that I am writing documentation for my future self as I frequently forget otherwise. Not everyone believes in in-line documentation but personally I think it is a must. 

#### 3. Handle Regex outside main function
If this function is called repeatedly, re-compiling the Regex every time can be inefficient. You could consider making the Regex static or lazily initialised to improve performance. For example, we can change:

```rust
pub fn mul_results_part_1(input: &str) -> usize {
    let re = Regex::new(r"mul\((\d{1,3}),(\d{1,3})\)").expect("Invalid regex");
    ...
}
```

to:

```rust
use regex::Regex;
use std::sync::LazyLock;

static MUL_REGEX: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"mul\((\d{1,3}),(\d{1,3})\)").expect("Invalid regex")
});

pub fn mul_results_part_1(input: &str) -> usize {
...
```
Lazy just means that it doesn't do anything until the first time it is needed. By putting the Regex expression outside the function, this means it will happen just once. 

Hope that helps! Well done :-)

**Comment by Sowmiya07** [comment]
[Here](https://github.com/Sowmiya07/AdventOfCode_WIR/blob/main/day_03/src/main.rs) is my Day 3 solution. 
I might've over-complicated it a bit... After checking out the solutions from others, it seems like I went for the complicated route.

Looks like it's time to finally face my fear of regex :p

I’m just glad I survived Day 3 with some new learnings.

**Reply by CarolineMorton** [reply]
I use ChatGPT to figure out Regex and then check on https://regex101.com! Don't think i have ever managed to write it by hand

**Reply by CarolineMorton** [reply]
Great test by the way 🎉🎉🎉

```rust
#[cfg(test)]
mod tests {
    use crate::multiplication;

    #[test]
    fn test_multiplication() {
        let corrupted_memory = "xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))";
        assert_eq!(multiplication(corrupted_memory), 161);
    }
}
```

**Comment by DominikStefancik** [comment]
My [solution for the day 3](https://github.com/DominikStefancik/AdventOfCode_WIR/tree/main/day_03). 
The second part took me way too much time. Maybe I overcomplicated the solution, but didn't come up with a better one.

**Comment by kuhmil** [comment]
Why did the Rust developer avoid using regex at work?
Because every time they tried, it matched.......their frustration!

Here is my [Day 3](https://github.com/kuhmil/AdventOfCode_WIR/tree/main/day_03)!

**Reply by CarolineMorton** [reply]
I really like how you have pulled the Regex statements out into their own Constants files. Stealing this idea for the future. It makes it a lot easier to read. 

https://github.com/kuhmil/AdventOfCode_WIR/blob/main/day_03/src/constants.rs
